src/gridLayer.ts

```diff
@@ -0,0 +1,186 @@
+      const newWidth = 4 * Math.ceil(width / 4);
+      return {
+        data: tile.data.map(data => padTileWithZeros({ data, width, height}, newWidth, height) as TypedArray),
+        height,
+        width: newWidth,
+      }
+    }
+    return tile;
+  };
+  return pMap(loaders, mapper, { concurrency });
+}
+
+export default class GridLayer<D, P extends GridLayerProps<D> = GridLayerProps<D>> extends CompositeLayer<D, P> {
+  initializeState() {
+    this.state = { gridData: [], width: 0, height: 0 };
+    refreshGridData(this.props).then(gridData => {
+      const { width, height } = validateWidthHeight(gridData);
+      this.setState({ gridData, width, height })
+    });
+  }
+
+  updateState({ props, oldProps, changeFlags }: { props: GridLayerProps<D>, oldProps: GridLayerProps<D>, changeFlags: any }) {
+    const { propsChanged } = changeFlags;
+    const loaderChanged = typeof propsChanged === 'string' && propsChanged.includes('props.loaders');
+    const loaderSelectionChanged = props.loaderSelection !== oldProps.loaderSelection;
+    if (loaderChanged || loaderSelectionChanged) {
+      // Only fetch new data to render if loader has changed
+      refreshGridData(this.props).then(gridData => {
+        this.setState({ gridData });
+      });
+    }
+  }
+
+  getPickingInfo({ info }: { info: any }) {
+    // provide Grid row and column info for mouse events (hover & click)
+    if (!info.coordinate) {
+      return info;
+    }
+    const { rows, columns } = this.props;
+    const spacer = this.props.spacer || 0;
+    const { width, height } = this.state;
+    const gridWidth = (columns * width) + ((columns - 1) * spacer);
+    const gridHeight = (rows * height) + ((rows - 1) * spacer);
+    const gridX = info.coordinate[0] + (gridWidth / 2);
+    const gridY = info.coordinate[1] + (gridHeight / 2);
+    const row = Math.floor(gridY / (height + spacer));
+    const column = Math.floor(gridX / (width + spacer));
+    info.gridCoord = { row, column }; // add custom property
+    return info;
+  }
+
+  renderLayers() {
+    const { gridData, width, height } = this.state;
+    if (width === 0 || height === 0) return null; // early return if no data
+
+    const { loaders, rows, columns, spacer, id = '' } = this.props;
+    const top = -(rows * (height + spacer)) / 2;
+    const left = -(columns * (width + spacer)) / 2;
+    const gridLayers = range(rows).flatMap((row) => {
+      return range(columns).map((col) => {
+        const y = top + (row * (height + spacer));
+        const x = left + (col * (width + spacer));
+        const offset = col + (row * columns);
+        const layerProps = {
+          channelData: gridData[offset] || null, // coerce to null if no data 
+          bounds: scaleBounds(width, height, [x, y]),
+          id: `${id}-GridLayer-${row}-${col}`,
+          dtype: loaders[offset]?.dtype || '<u2', // fallback if missing,
+        };
+        return new XRLayer({ ...this.props, ...layerProps });
```

Comment:
I know I'm a little late to the party but I was using the demo last night and noticed that it takes a while for the whole thing to load - was there a reason not to use `ImageLayer` so that each small image loads independently?  I might have missed something in the course of this PR that would prevent doing that.

Reply:
Ok, maybe a future thing would be to wrap the loaders then so that the new `getTile` function calls the old one with a concurrency limit.  Future improvements :)